unified-codes/apps/data-loader/src/app/v2/DataParser.ts

```diff
@@ -200,403 +185,83 @@ export class DataParser {
-              );
+            if (item.type && this.graph[item.code].type != item.type) {
+              duplicates.push(item.code);
             }
           }
-        }
+        });
 
-        // If unit code exists...
-        if (uc7) {
-          // and pack immediate code exists...
-          if (uc8) {
-            // link unit to pack immediate.
-            if (!this.graph[uc7].children.map((child) => child.code).includes(uc8)) {
-              this.graph[uc7].children.push({ code: uc8 });
-              console.log(`INFO: Linked unit with code ${uc7} to pack immediate with code ${uc8}`);
-            }
-          }
-        }
+        const productCode = productDefinition[0].code;
+        // If category code exists, and product code exists
+        if (uc1 && productCode) {
 
-        // If dose strength code exists...
-        if (uc6) {
-          // and unit code exists...
-          if (uc7) {
-            // link dose strength to unit.
-            if (!this.graph[uc6].children.map((child) => child.code).includes(uc7)) {
-              this.graph[uc6].children.push({ code: uc7 });
-              console.log(`INFO: Linked dose strength with code ${uc6} to unit with code ${uc7}`);
-            }
+          // link category to product.
+          if (!this.graph[uc1].children.map((child) => child.code).includes(productCode)) {
+            this.graph[uc1].children.push({ code: productCode });
+            console.log(`INFO: Linked category with code ${uc1} to product with code ${productCode}`);
           }
         }
 
-        // If dose qualifier code exists...
-        if (uc5) {
-          // and strength code exists...
-          if (uc6) {
-            // link dose qualification to strength.
-            if (!this.graph[uc5].children.map((child) => child.code).includes(uc6)) {
-              this.graph[uc5].children.push({ code: uc6 });
+        // Iterate through and assign children to parents 
+        // This starts at the top of the graph and for each node, finds the closest lower level node that has a code and name
+        // This avoids having to manually define each combination of 'levels' that are able to be connected 
+        let parentIndex = 0;
+        let childIndex = 1;
+        console.log(productDefinition)
+        while (childIndex < productDefinition.length) {
+          const parent = productDefinition[parentIndex];
+          const child = productDefinition[childIndex];
+          if (child.name && child.code) {
+            if (!this.graph[parent.code].children.map((child) => child.code).includes(child.code)) {
+              this.graph[parent.code].children.push(child);
               console.log(
-                `INFO: Linked dose qualifier with code ${uc5} to dose strength with code ${uc6}`
-              );
-            }
-          }
-          // if strength doesn't exist but unit of presentation does
-          else if (uc7) {
```

Comment:
I'm not seeing where this has gone in the new version

Reply:
The idea was that this would / should handle each case of checking a node for its child:
https://github.com/openmsupply/unified-codes/blob/b4377e7168ebf83baaf8fcbd33360082a5c84cf5/unified-codes/apps/data-loader/src/app/v2/DataParser.ts#L225-L240

E.g if UC2-7 were all defined except for UC6 (strength) this block should do the following linkages  
- UC2 -> UC3 (`productDefinition[0]` and `productDefinition[1]`)
- UC3 -> UC4 (`productDefinition[1]` and `productDefinition[2]`)
- UC4 -> UC5 (`productDefinition[2]` and `productDefinition[3]`)
- `// UC6 skipped because there's no name and code defined for productDefinition[4]`
- UC5 -> UC7 (`productDefinition[3]` and `productDefinition[5]`)

... Hmm I'm not sure I'm explaining very well (I guess a unit test would also be good for a function like this! 🙀

Reply:
We also have these cases in the current version which don't map to anything in the new that I can see:
- `uc2 && !uc3 && uc6` 
- `uc2 && !uc3 && !uc6 && uc7`
- `uc5 && !uc6 && uc7`

Since they've deliberately been put in place, I think we should reimplement, sorry!

Reply:
Hmm, going to need a re-think on the algorithm, perhaps. What I've done I guess is a bit blind to the structure validation rules and relies on accurate spreadsheet entry, whereas the previous implementation has built-in assumptions about data structure rules.

With the existing logic, all of the above scenarios will be covered, but only if there aren't any codes in between.

E.g. For the first example
uc2 will still be mapped to uc6 if there is no 3, 4 or 5 defined.
If there is a uc4 (someone put it in the spreadsheet by accident), then the structure created will be uc2 -> uc4 -> uc6 which is technically invalid according to the SoP: 
- Items cannot have a dose form without a route:  https://docs.google.com/document/d/1xKmWT0Qiyam7tHBUTjZUteyF8VULOYFG/edit)  😢 

Maaaaybe, I should revert the tree builder and leave the old logic, it just hurts so much to read and it also skips nodes without warning 😆. Perhaps a better solution would be to log the skipped codes somewhere?